File: libraries/node/src/parser/checks.ts
import { UrcError } from "../error.js";
import {
	METADATA_FIELDS,
	SECTION_JUDGMENT,
	SECTION_LAYOUT,
	SECTION_METADATA,
	SECTION_TIMING,
	rule
} from "../strings.js";
import type { ParseState } from "./state.js";

/**
 * Runs the finalize-time checks for a closed section.
 */
export function finalizeSection(state: ParseState, section: string, line: number): void {
	if (section === SECTION_METADATA) 
		checkMetadataComplete(state, line);
	else if (section === SECTION_JUDGMENT) 
		checkJudgment(state, line);
	else if (section === SECTION_LAYOUT) 
		checkLayout(state, line);
	else if (section === SECTION_TIMING && state.timing.length === 0) 
		throw new UrcError(rule(14), line, "first timing point must be at timestamp 0");
	
}

function checkMetadataComplete(state: ParseState, line: number): void {
	for (const name of METADATA_FIELDS) 
		if (!state.metadata.has(name)) 
			throw new UrcError(rule(4), line, `Metadata is missing field: ${name}`);
	
}

function checkJudgment(state: ParseState, line: number): void {
	const { windows, rates } = state;
	if (windows === null || rates === null) 
		throw new UrcError(rule(4), line, "Judgment requires both Window and Rate");
	
	if (windows.length !== rates.length) 
		throw new UrcError(rule(7), line, "Window and Rate must have the same count");
	
	for (let index = 1; index < windows.length; index++) 
		if (windows[index] < windows[index - 1]) 
			throw new UrcError(rule(8), line, "Window values must be ascending");
	
	for (let index = 1; index < rates.length; index++) 
		if (rates[index] > rates[index - 1]) 
			throw new UrcError(rule(9), line, "Rate values must be descending");
	
	for (const rate of rates) 
		if (rate < 0 || rate > 100) 
			throw new UrcError(rule(10), line, "Rate values must be in 0-100");
	
}

function checkLayout(state: ParseState, line: number): void {
	const { layoutType, special, specialSeen } = state;
	if (layoutType === null) 
		throw new UrcError(rule(4), line, "Layout is missing field: Type");
	
	if (!specialSeen) 
		throw new UrcError(rule(4), line, "Layout is missing field: Special");
	
	if (special === null) 
		return;
	
	const total = layoutType[0] + layoutType[1];
	for (const lane of special) 
		if (lane < 0 || lane >= total) 
			throw new UrcError(rule(12), line, `special lane out of range: ${lane}`);
	
	if (new Set(special).size !== special.length) 
		throw new UrcError(rule(13), line, "duplicate special lanes");
	
}

File: libraries/rust/src/parser/checks.rs
//! Finalize-time semantic checks (Phase A completion and Phase B).

use super::state::ParseState;
use crate::error::{Result, UrcError};
use crate::strings::{
    METADATA_FIELDS, SECTION_JUDGMENT, SECTION_LAYOUT, SECTION_METADATA, SECTION_TIMING, rule,
};

pub(super) fn finalize_section(state: &mut ParseState, section: &str, line: u32) -> Result<()> {
    match section {
        SECTION_METADATA => check_metadata_complete(state, line),
        SECTION_JUDGMENT => check_judgment(state, line),
        SECTION_LAYOUT => check_layout(state, line),
        SECTION_TIMING if state.timing.is_empty() => Err(UrcError::new(
            rule(14),
            line,
            "first timing point must be at timestamp 0",
        )),
        _ => Ok(()),
    }
}

fn check_metadata_complete(state: &mut ParseState, line: u32) -> Result<()> {
    for field in METADATA_FIELDS {
        if !state.metadata.iter().any(|(name, _)| name == field) {
            return Err(UrcError::new(
                rule(4),
                line,
                format!("Metadata is missing field: {field}"),
            ));
        }
    }
    Ok(())
}

fn check_judgment(state: &mut ParseState, line: u32) -> Result<()> {
    let (Some(windows), Some(rates)) = (&state.windows, &state.rates) else {
        return Err(UrcError::new(
            rule(4),
            line,
            "Judgment requires both Window and Rate",
        ));
    };

    if windows.len() != rates.len() {
        return Err(UrcError::new(
            rule(7),
            line,
            "Window and Rate must have the same count",
        ));
    }
    if windows.windows(2).any(|pair| pair[1] < pair[0]) {
        return Err(UrcError::new(
            rule(8),
            line,
            "Window values must be ascending",
        ));
    }
    if rates.windows(2).any(|pair| pair[1] > pair[0]) {
        return Err(UrcError::new(
            rule(9),
            line,
            "Rate values must be descending",
        ));
    }
    if rates.iter().any(|&rate| !(0.0..=100.0).contains(&rate)) {
        return Err(UrcError::new(
            rule(10),
            line,
            "Rate values must be in 0-100",
        ));
    }
    Ok(())
}

fn check_layout(state: &mut ParseState, line: u32) -> Result<()> {
    let Some((keys, special_keys)) = state.layout_type else {
        return Err(UrcError::new(
            rule(4),
            line,
            "Layout is missing field: Type",
        ));
    };
    if !state.special_seen {
        return Err(UrcError::new(
            rule(4),
            line,
            "Layout is missing field: Special",
        ));
    }

    let Some(lanes) = &state.special else {
        return Ok(());
    };
    let total = keys + special_keys;
    for &lane in lanes {
        if lane < 0 || lane as u64 >= total {
            return Err(UrcError::new(
                rule(12),
                line,
                format!("special lane out of range: {lane}"),
            ));
        }
    }

    let mut unique = lanes.clone();
    unique.sort_unstable();
    unique.dedup();
    if unique.len() != lanes.len() {
        return Err(UrcError::new(rule(13), line, "duplicate special lanes"));
    }
    Ok(())
}
